File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/Presentation.java
package it.unibo.coordination.linda.presentation;

import alice.tuprolog.Term;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import it.unibo.coordination.linda.core.TupleSpace;
import it.unibo.coordination.linda.core.events.TupleSpaceEvent;
import it.unibo.coordination.linda.logic.InspectableLogicSpace;
import it.unibo.coordination.linda.logic.LogicMatch;
import it.unibo.coordination.linda.logic.LogicTemplate;
import it.unibo.coordination.linda.logic.LogicTuple;
import it.unibo.coordination.linda.string.RegexTemplate;
import it.unibo.coordination.linda.string.RegularMatch;
import it.unibo.coordination.linda.string.StringTuple;
import it.unibo.coordination.prologx.PrologUtils;
import org.jooq.lambda.function.Function2;
import org.jooq.lambda.function.Function3;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.lang.reflect.InvocationTargetException;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

public class Presentation {

    static {
    }

    private static final Map<Tuple2<Class<?>, MIMETypes>, Serializer<?>> serializers = new HashMap<>();
    private static final Map<Tuple2<Class<?>, MIMETypes>, Deserializer<?>> deserializers = new HashMap<>();

    private static final Map<MIMETypes, ObjectMapper> mappers = Map.of(
        MIMETypes.APPLICATION_JSON, createMapper(ObjectMapper.class),
        MIMETypes.APPLICATION_YAML, createMapper(YAMLMapper.class),
        MIMETypes.APPLICATION_XML, createMapper(XmlMapper.class)
    );

    public static <OM extends ObjectMapper> OM createMapper(Class<OM> mapperClass) {
        try {
            final OM mapper = mapperClass.getConstructor().newInstance();
            mapper.registerModule(new JavaTimeModule());
            return mapper;
        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new IllegalStateException(e);
        }
    }

    public static <T> Serializer<T> getSerializer(Class<T> type, MIMETypes mimeType) {
        var result = serializers.get(Tuple.tuple(type, mimeType));
        if (result == null) {
            throw new IllegalArgumentException("Class-MIMEType combo not supported: " + type.getName() + " --> " + mimeType);
        }
        return (Serializer<T>) result;
    }

    public static <T> Deserializer<T> getDeserializer(Class<T> type, MIMETypes mimeType) {
        var result = deserializers.get(Tuple.tuple(type, mimeType));
        if (result == null) {
            throw new IllegalArgumentException("Class-MIMEType combo not supported: " + type.getName() + " <-- " + mimeType);
        }
        return (Deserializer<T>) result;
    }

    public static <T> void register(Class<T> type, Serializer<? super T> serializer) {
        final Tuple2<Class<?>, MIMETypes> key = Tuple.tuple(type, serializer.getSupportedMIMEType());
        if (serializers.containsKey(key)) {
            throw new IllegalArgumentException("Class-MIMEType combo already registered: " + type.getName() + " --> " + serializer.getSupportedMIMEType());
        }
        serializers.put(key, serializer);
    }

    public static <T> void registerSimpleSerializers(Class<T> type) {
        registerSimpleSerializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML));
    }

    public static <T> void registerSimpleSerializers(Class<T> type, EnumSet<MIMETypes> types) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, new SimpleSerializer<>(t, mapper));
        }
    }

    public static <T> void registerDynamicSerializers(Class<T> type, Function3<Class<T>, MIMETypes, T, Object> f) {
        registerDynamicSerializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML), f);
    }

    public static <T> void registerDynamicSerializers(Class<T> type, EnumSet<MIMETypes> types, Function3<Class<T>, MIMETypes, T, Object> f) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, new DynamicSerializer<T>(t, mapper) {
                @Override
                public Object toDynamicObject(T object) {
                    return f.apply(type, getSupportedMIMEType(), object);
                }
            });
        }
    }

    public static <T> void registerDynamicSerializers(Class<T> type, Function2<MIMETypes, ObjectMapper, DynamicSerializer<T>> f) {
        registerDynamicSerializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML), f);
    }

    public static <T> void registerDynamicSerializers(Class<T> type, EnumSet<MIMETypes> types, Function2<MIMETypes, ObjectMapper, DynamicSerializer<T>> f) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, f.apply(t, mapper));
        }
    }

    public static <T> void register(Class<T> type, Deserializer<? extends T> deserializer) {
        final Tuple2<Class<?>, MIMETypes> key = Tuple.tuple(type, deserializer.getSupportedMIMEType());
        if (deserializers.containsKey(key)) {
            throw new IllegalArgumentException("Class-MIMEType combo already registered: " + type.getName() + " <-- " + deserializer.getSupportedMIMEType());
        }
        deserializers.put(key, deserializer);
    }

    public static <T> void registerSimpleDeserializers(Class<T> type) {
        registerSimpleDeserializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML));
    }

    public static <T> void registerSimpleDeserializers(Class<T> type, EnumSet<MIMETypes> types) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, new SimpleDeserializer<>(type, t, mapper));
        }
    }

    public static <T> void registerDynamicDeserializers(Class<T> type, Function3<Class<T>, MIMETypes, Object, T> f) {
        registerDynamicDeserializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML), f);
    }

    public static <T> void registerDynamicDeserializers(Class<T> type, EnumSet<MIMETypes> types, Function3<Class<T>, MIMETypes, Object, T> f) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, new DynamicDeserializer<T>(type, t, mapper) {

                @Override
                public T fromDynamicObject(Object dynamicObject) {
                    return f.apply(type, t, dynamicObject);
                }
            });
        }
    }

    public static <T> void registerDynamicDeserializers(Class<T> type, Function2<MIMETypes, ObjectMapper, DynamicDeserializer<T>> f) {
        registerDynamicDeserializers(type, EnumSet.of(MIMETypes.APPLICATION_JSON, MIMETypes.APPLICATION_YAML, MIMETypes.APPLICATION_XML), f);
    }

    public static <T> void registerDynamicDeserializers(Class<T> type, EnumSet<MIMETypes> types, Function2<MIMETypes, ObjectMapper, DynamicDeserializer<T>> f) {
        for (var t : types) {
            final var mapper = mappers.get(t);
            if (mapper == null) throw new IllegalArgumentException("No mapper for MIMEType " + t);
            register(type, f.apply(t, mapper));
        }
    }

    static {
        registerDynamicSerializers(LogicTemplate.class, (klass, targetType, template) ->
                PrologUtils.termToDynamicObject(template.getTemplate()));
        registerDynamicSerializers(LogicTuple.class, (klass, targetType, tuple) ->
                PrologUtils.termToDynamicObject(tuple.getValue()));
        registerDynamicSerializers(LogicMatch.class, LogicMatchSerializer::new);
        registerDynamicSerializers(Term.class, (klass, targetType, term) ->
                PrologUtils.termToDynamicObject(term));
        registerDynamicSerializers(TupleSpaceEvent.class, (mimeType, mapper) -> new TupleSpaceEventSerializer(mimeType, mapper, LogicTuple.class, LogicTemplate.class));

        registerDynamicDeserializers(LogicTemplate.class, (klass, targetType, obj) ->
                LogicTemplate.of(PrologUtils.dynamicObjectToTerm(obj)));
        registerDynamicDeserializers(LogicTuple.class, (klass, targetType, obj) ->
                LogicTuple.of(PrologUtils.dynamicObjectToTerm(obj)));
        registerDynamicDeserializers(LogicMatch.class, LogicMatchDeserializer::new);
        registerDynamicDeserializers(Term.class, (klass, targetType, obj) ->
                PrologUtils.dynamicObjectToTerm(obj));
        registerDynamicDeserializers(TupleSpaceEvent.class, (mimeType, mapper) -> new TupleSpaceEventDeserializer(mimeType, mapper, LogicTuple.class, LogicTemplate.class) {
            @Override
            protected TupleSpace getSpace(String tupleSpaceName) {
                return InspectableLogicSpace.create(tupleSpaceName);
            }
        });


        registerDynamicSerializers(StringTuple.class, StringTupleSerializer::new);
        registerDynamicSerializers(RegexTemplate.class, RegexTemplateSerializer::new);
        registerDynamicSerializers(RegularMatch.class, RegularMatchSerializer::new);

        registerDynamicDeserializers(StringTuple.class, StringTupleDeserializer::new);
        registerDynamicDeserializers(RegexTemplate.class, RegexTemplateDeserializer::new);
        registerDynamicDeserializers(RegularMatch.class, RegularMatchDeserializer::new);
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/Presentation.kt
package it.unibo.presentation

import com.fasterxml.jackson.databind.ObjectMapper
import java.util.*

interface Presentation {
    @JvmDefault
    fun <T> register(type: Class<T>, serializer: Serializer<T>) {
        register(type.toTypeToken(), serializer)
    }

    fun <T> register(typeToken: TypeToken<T>, serializer: Serializer<T>)

    @JvmDefault
    fun <T> register(type: Class<T>, deserializer: Deserializer<T>) {
        register(type.toTypeToken(), deserializer)
    }
    fun <T> register(typeToken: TypeToken<T>, deserializer: Deserializer<T>)

    @JvmDefault
    fun <T> serializerOf(type: Class<T>, mimeType: MIMETypes): Serializer<T> =
            serializerOf(type.toTypeToken(), mimeType)

    fun <T> serializerOf(typeToken: TypeToken<T>, mimeType: MIMETypes): Serializer<T>

    @JvmDefault
    fun <T> deserializerOf(type: Class<T>, mimeType: MIMETypes): Deserializer<T> =
            deserializerOf(type.toTypeToken(), mimeType)

    fun <T> deserializerOf(typeToken: TypeToken<T>, mimeType: MIMETypes): Deserializer<T>

    @JvmDefault
    fun <T> registerSimpleSerializers(type: Class<T>) =
            registerSimpleSerializers(type.toTypeToken())

    @JvmDefault
    fun <T> registerSimpleSerializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>) =
            registerSimpleSerializers(type.toTypeToken(), mimeTypes)

    @JvmDefault
    fun <T> registerDynamicSerializers(type: Class<T>, f: (MIMETypes, ObjectMapper) -> Serializer<T>) =
            registerDynamicSerializers(type.toTypeToken(), f)

    @JvmDefault
    fun <T> registerDynamicSerializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>, f: (MIMETypes, ObjectMapper) -> Serializer<T>) =
            registerDynamicSerializers(type.toTypeToken(), mimeTypes, f)

    @JvmDefault
    fun <T> registerDynamicSerializers(type: Class<T>, f: (Class<T>, ObjectMapper, T) -> Any) =
            registerDynamicSerializers(type.toTypeToken()) { token, mapper, obj ->
                f(token.type, mapper, obj)
            }

    @JvmDefault
    fun <T> registerDynamicSerializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>, f: (Class<T>, ObjectMapper, T) -> Any) =
            registerDynamicSerializers(type.toTypeToken(), mimeTypes) { token, mapper, obj ->
                f(token.type, mapper, obj)
            }

    @JvmDefault
    fun <T> registerSimpleDeserializers(type: Class<T>) =
            registerSimpleDeserializers(type.toTypeToken())

    @JvmDefault
    fun <T> registerSimpleDeserializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>) =
            registerSimpleDeserializers(type.toTypeToken(), mimeTypes)

    @JvmDefault
    fun <T> registerDynamicDeserializers(type: Class<T>, f: (Class<T>, ObjectMapper, Any) -> T) =
            registerDynamicDeserializers(type.toTypeToken()) { token, mapper, obj ->
                f(token.type, mapper, obj)
            }

    @JvmDefault
    fun <T> registerDynamicDeserializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>, f: (Class<T>, ObjectMapper, Any) -> T) =
            registerDynamicDeserializers(type.toTypeToken(), mimeTypes) { token, mapper, obj ->
                f(token.type, mapper, obj)
            }

    @JvmDefault
    fun <T> registerDynamicDeserializers(type: Class<T>, f: (MIMETypes, ObjectMapper) -> Deserializer<T>) =
            registerDynamicDeserializers(type.toTypeToken(), f)

    @JvmDefault
    fun <T> registerDynamicDeserializers(type: Class<T>, mimeTypes: EnumSet<MIMETypes>, f: (MIMETypes, ObjectMapper) -> Deserializer<T>) =
            registerDynamicDeserializers(type.toTypeToken(), mimeTypes, f)

    fun <T> registerSimpleSerializers(typeToken: TypeToken<T>)
    fun <T> registerSimpleSerializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>)
    fun <T> registerDynamicSerializers(typeToken: TypeToken<T>, f: (MIMETypes, ObjectMapper) -> Serializer<T>)
    fun <T> registerDynamicSerializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>, f: (MIMETypes, ObjectMapper) -> Serializer<T>)
    fun <T> registerDynamicSerializers(typeToken: TypeToken<T>, f: (TypeToken<T>, ObjectMapper, T) -> Any)
    fun <T> registerDynamicSerializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>, f: (TypeToken<T>, ObjectMapper, T) -> Any)

    fun <T> registerSimpleDeserializers(typeToken: TypeToken<T>)
    fun <T> registerSimpleDeserializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>)
    fun <T> registerDynamicDeserializers(typeToken: TypeToken<T>, f: (TypeToken<T>, ObjectMapper, Any) -> T)
    fun <T> registerDynamicDeserializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>, f: (TypeToken<T>, ObjectMapper, Any) -> T)
    fun <T> registerDynamicDeserializers(typeToken: TypeToken<T>, f: (MIMETypes, ObjectMapper) -> Deserializer<T>)
    fun <T> registerDynamicDeserializers(typeToken: TypeToken<T>, mimeTypes: EnumSet<MIMETypes>, f: (MIMETypes, ObjectMapper) -> Deserializer<T>)

    fun <M : ObjectMapper> getMapper(mimeType: MIMETypes): M

    companion object Prototype : Presentation by PresentationImpl()
}